File: 0622. Design Circular Queue.java
// Medium
// Queue, Design
// Time:O(1),Space:O(k)
// https://leetcode.cn/problems/design-circular-queue/

class MyCircularQueue {
    private int[] que;
    private int front, rear, size;

    public MyCircularQueue(int k) {
        front = 0;
        rear = 0;
        size = k + 1;
        que = new int[size];
    }
    
    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }
        que[rear] = value;
        rear = (rear + 1) % size; // update rear pointer
        return true;
    }
    
    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }
        front = (front + 1) % size; // update front pointer
        return true;
    }
    
    public int Front() {
        if (isEmpty()) {
            return -1;
        }
        return que[front];
    }
    
    public int Rear() {
        if (isEmpty()) {
            return -1;
        }
        return que[(rear - 1 + size) % size]; // deal with circular
    }
    
    public boolean isEmpty() {
        return front == rear;
    }
    
    public boolean isFull() {
        return (rear + 1) % size == front;
    }
}

File: 0501-1000/0622. Design Circular Queue.py
# Medium
# Design
# Time:O(mnk), Space:O(mnk)
# https://leetcode.cn/problems/design-circular-queue/

class MyCircularQueue:

    def __init__(self, k: int):
        self.capacity = k
        self.que = [0] * k # fixed size
        self.front = 0 # front pointer
        self.rear = -1 # rear pointer
        self.size = 0 # cur num of elements

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False
        
        # latest: move the rear pointer
        # mod capacity for circular
        self.rear = (self.rear + 1) % self.capacity
        self.que[self.rear] = value
        self.size += 1
        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False
        
        # move the front pointer
        # to del the earliest element
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return True

    def Front(self) -> int:
        if self.isEmpty():
            return -1
        
        return self.que[self.front]

    def Rear(self) -> int:
        if self.isEmpty():
            return -1
        
        return self.que[self.rear]

    def isEmpty(self) -> bool:
        return self.size == 0
    
    def isFull(self) -> bool:
        return self.size == self.capacity
    